src/ripple/core/impl/Workers.cpp

```diff
@@ -17,264 +17,113 @@
-                //
-                worker->notify();
-            }
-            else
-            {
-                worker = new Worker(*this, m_threadNames, instance++);
-                m_everyone.push_front(worker);
-            }
-        }
-    }
-    else
-    {
-        // Decreasing the number of working threads
-        int const amount = m_numberOfThreads - numberOfThreads;
+                threads_++;
 
-        for (int i = 0; i < amount; ++i)
-        {
-            ++m_pauseCount;
-
-            // Pausing a thread counts as one "internal task"
-            m_semaphore.notify();
-        }
-    }
-
-    m_numberOfThreads = numberOfThreads;
-}
-
-void
-Workers::stop()
-{
-    setNumberOfThreads(0);
-
-    std::unique_lock<std::mutex> lk{m_mut};
-    m_cv.wait(lk, [this] { return m_allPaused; });
-    lk.unlock();
+                while (!stopping_)
+                {
+                    auto t = tail_.load();
+                    auto h = head_.load();
+
+                    assert(h >= t);
+
+                    beast::setCurrentThreadName(n + " [zZz]");
+
+                    // If there's nothing to do, go to sleep
+                    while (h == t && !stopping_.load(std::memory_order_relaxed))
+                    {
+                        paused_++;
+                        {
+                            std::unique_lock lock(m_mut);
+                            m_cv.wait_for(lock, std::chrono::seconds(5));
+                        }
+                        t = tail_.load();
+                        h = head_.load();
+                        paused_--;
+                    }
+
+                    if (!stopping_.load(std::memory_order_relaxed) && h > t)
+                    {
+                        // As long as we aren't stopping and there is a task
+                        // that's waiting, we try to do work.
+                        if (tail_.compare_exchange_strong(t, t + 1))
+                        {
+                            // Restore the name, in case the previous callback
+                            // had changed it.
+                            beast::setCurrentThreadName(n);
+
+                            // Exceptions should never bubble up to here but
+                            // just in case, if one does catch it.
+                            try
+                            {
+                                m_callback.processTask(instance);
+                            }
+                            catch (...)
+                            {
+                                m_callback.uncaughtException(
+                                    instance, std::current_exception());
+                            }
+                        }
+                    }
+                }
 
-    assert(numberOfCurrentlyRunningTasks() == 0);
+                // Track number of threads
+                threads_--;
+            },
+            count);
+        t.detach();
```

Comment:
I'd probably keep a collection of threads in the object rather than `detach` here.

Reply:
@scottschurr, @seelabs: I thought about this and actually wrote the version with the `std::vector<std::thread>` and _something_ about it rubbed me the wrong way. I guess it's because actually holding onto the handle doesn't really serve a purpose.

Unless you're vetoing the current version (and if you feel strongly about it, please do!), I'd prefer to stick with the version as is.